Sources/AsyncHTTPClient/HTTPClient.swift

```diff
@@ -54,6 +75,8 @@ public class HTTPClient {
     var state: State
     private let stateLock = Lock()
 
+    public static let loggingDisabled = Logger(label: "AHC-do-not-log", factory: { _ in DoNotLogger() })
```

Comment:
why do we need this? Why not just a new `Logger` and let clients decide if they want to log stuff?

Reply:
Well, by default I think a HTTP client shouldn't log. I added some more info here: https://github.com/swift-server/async-http-client/pull/227/files#diff-e92f784d7349c391e323afad21d5702e

Reply:
[the code doesn't reflect the design yet :P ]

Reply:
This design is a bit weird for me. Usually, as a client of a library, I configure logging using levels and configuration. That allows to configure all dependencies in one way and in one place. With this design I will have to modify my code if I want to log. Having default level for different components allows to turn on logging with a config and a restart

Reply:
@artemredkin no, you don't have to modify your code. You just hand AHC _your logger_ and it'll log on that one. The `loggingDisabled` is just the default if you do not hand us a logger.

Reply:
If my program compiled without passing logger to every invocation - then I have to recompile, no? I would prefer for this to be a configuration thing, not a custom logging backend thing. This pattern is unlike anything I've seen so far.

Reply:
@artemredkin You will _have_ to pass a logger in any system that you want to log if you're using structured logging. If you don't explicitly pass loggers, there's no way we can preserve  your metadata like your request ID.

Let's say you have a system that does

```
logger[metadataKey: "request-id"] = "really cool request"
httpClient.get("https://...", logger: logger)
httpClient.post("https://...", logger: logger)
```

then you want your "really cool request" request ID be attached to all messages logged throughout the program that have anything to do with your request. AHC must accept a `Logger` in order to get hold of the metadata you supply.

Reply:
See here for the unit test that makes sure this works: https://github.com/swift-server/async-http-client/pull/227/files#diff-bd03a2521c79f3da4bd5825512ca9979R1761-R1780

Reply:
In RediStack I log by default with my own metadata and label. If you pass me a logger, then I will use your logger with my own metadata being attached to it with the label you gave it.

Right now I'm leaving it up to Logging backends to provide a way to disable RediStack as a system

Reply:
@Mordil but in RediStack at the moment you set the the logging [on the client](https://gitlab.com/Mordil/RediStack/-/blob/1.0.0-alpha.9/Sources/RediStack/RedisClient.swift#L41). That won't work if you have two concurrent requests or am I missing something.

As an example

```
thread1 { logger in
    let requestID = UUID()
    var logger = logger
    logger["my-req-id"] = "\(requestID)"
    
    redisClient.setLogger(to: logger)
    redisClient.send(command: foo)
}

thread2 { logger in
    let requestID = UUID()
    var logger = logger
    logger["my-req-id"] = "\(requestID)"
    
    redisClient.setLogger(to: logger)
    redisClient.send(command: bar)
}
```

Now, probably the metadata is actually mixed up if thread1 and 2 run concurrently.

I think you'll need to switch to passing the logger with the request, like so

```
thread1 { logger in
    let requestID = UUID()
    var logger = logger
    logger["my-req-id"] = "\(requestID)"
    
    redisClient.send(command: foo, logger: logger)
}

thread2 { logger in
    let requestID = UUID()
    var logger = logger
    logger["my-req-id"] = "\(requestID)"
    
    redisClient.send(command: bar, logger: logger)
}
```

or am I missing somethign?

Reply:
Yes, the logging implementation is a bit naïve, as I was sloppy and rushed an implementation that solved a use case for Vapor that matched how Vapor wanted to do things (at the time).

I have this bug against RediStack that I'm going to use for addressing it - but I was going to wait until seeing the AHC implementation go through so I can take notes from it: https://gitlab.com/Mordil/RediStack/-/issues/79

The point I was making is that my intention is RediStack _always_ logs, but the user can override the label and provide their own metadata that RS will attach 1 or 2 pieces to.